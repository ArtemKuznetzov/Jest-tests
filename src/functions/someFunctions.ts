import { v4 } from 'uuid';
import {passwordValidateErrors} from "../__mocks__/passwordValidateErrors";

export const sumNums = (a: number, b: number) => a + b;

export const divideNums = (a: number, b: number) => a / b;

export const multiplyNums = (a: number, b: number) => a * b;

export const getLength = <T, >(a: Array<T>) => a.length;

export const filterArray = <T, F extends Function>(array: Array<T>, callback: F) => {
    const newArray = [];

    for (let i = 0; i < array.length; i++) {
        console.log(array[i]);
        if (callback(array[i])) newArray.push(array[i]);
    }

    return newArray;
};

export const createTodo = (title: string) => {
    if (!title) {
        throw new Error('title is required')
    }
    return {
        title,
        completed: false,
        id: v4(),
    }};

// Раньше функция возвращала только true или false. Допустим пришли изменения от заказчика, функционал изменился.
export const validatePassword = (password: string) => {
    if (password.length < 8) {
        return {
            success: false,
            error: passwordValidateErrors.length
        }
    }

    if (!/[a-z]/.test(password) || !/[A-Z]/.test(password)) {
        return {
            success: false,
            error: passwordValidateErrors.case
        }
    }

    if (!/[0-9]/.test(password)) {
        {
            return {
                success: false,
                error: passwordValidateErrors.number
            }
        }
    }

    if (!/[^A-Za-z0-9]/.test(password)) {
        return {
            success: false,
            error: passwordValidateErrors.special
        }
    }

    return { success: true, error: null }
}

export const createUser = ({
    firstName, lastName, phone = '', email = ''
}: {firstName: string, lastName: string, phone?: string, email?: string}) => ({
  type: 'USER',
  firstName,
  lastName,
  password: 'password',
  phone,
  email
})

export const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

