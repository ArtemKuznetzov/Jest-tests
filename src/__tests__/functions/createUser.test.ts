import {createUser} from "../../functions/someFunctions";

describe('create user', () => {
    it('user with first name and last name only', () => {
        const user = createUser({
            firstName: 'Jhon',
            lastName: 'Stone'
        })
        // при первом запуске теста создастся папка __snapshots__ с снепшотом. При последующем
        // запуске теста jest будет сверяться с уже существующим. чтобы обновить данные в снепшоте
        // нужно поменять их вручную и запустить команду npm unit -- -u
        expect(user).toMatchSnapshot()
    })

    it('user with first, last name and email', () => {
        const user = createUser({
            firstName: 'Jhon',
            lastName: 'Stone',
            email: 'jhon.stone@mail.ru'
        })

        expect(user).toMatchSnapshot()
    })

    it('user with first, last name and phone', () => {
        const user = createUser({
            firstName: 'Jhon',
            lastName: 'Stone',
            phone: '+1-234-567-89-01'
        })

        expect(user).toMatchSnapshot()
    })
})