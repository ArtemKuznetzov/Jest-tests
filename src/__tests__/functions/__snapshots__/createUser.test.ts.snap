// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create user user with first name and last name only 1`] = `
{
  "email": "",
  "firstName": "Jhon",
  "lastName": "Stone",
  "password": "password",
  "phone": "",
  "type": "USER",
}
`;

exports[`create user user with first, last name and email 1`] = `
{
  "email": "jhon.stone@mail.ru",
  "firstName": "Jhon",
  "lastName": "Stone",
  "password": "password",
  "phone": "",
  "type": "USER",
}
`;

exports[`create user user with first, last name and phone 1`] = `
{
  "email": "",
  "firstName": "Jhon",
  "lastName": "Stone",
  "password": "password",
  "phone": "+1-234-567-89-01",
  "type": "USER",
}
`;
